book/chapters/appendices/solutions.qmd

```diff
@@ -800,72 +801,174 @@ system.time(resample(task, learner, resampling))
+        properties = "requires_task", # needs access to task for positive class
+        predict_type = "prob", # measures probability prediction
+        range = c(0, 1), # results in values between [0, 1]
         minimize = TRUE # smaller values are better
       )
     }
   ),
 
   private = list(
-    .score = function(prediction, ...) { # define score as private method
-      # define loss
-      costsens = function(truth, prediction) {
-        score = numeric(length(truth))
-        score[truth == "A" & prediction == "B"] = 10
-        score[truth == "B" & prediction == "A"] = 1
-
-        mean(score)
-      }
-
+    .score = function(prediction, task, ...) { # define score as private method
       # call loss function
-      costsens(prediction$truth, prediction$response)
+      mae_prob(prediction$truth, prediction$prob, task)
     }
   )
 )
 ```
 
-An alternative (as pointed to by the hint) can be constructed by first translating the rules to a matrix of misclassification costs, and then feeding this matrix to the constructor of `msr("classif.costs")`:
+Because this is a custom class that is not available in the `mlr_measures` dictionary, we have to create a new instance using the `$new()` constructor.
+
+```{r}
+msr_mae_prob = MeasureMaeProb$new()
+msr_mae_prob
+```
+
+
+To try this measure, we resample a logistic regression on the sonar task using five-fold cross-validation.
+
+```{r}
+learner = lrn("classif.log_reg", predict_type = "prob") # otherwise our meausre does not work
+task = tsk("sonar")
+rr = resample(task, learner, rsmp("cv", folds = 5))
+```
+
+We now score the resample result using our custom measure and `msr("classif.acc")`.
+
+```{r}
+score = rr$score(list(msr_mae_prob, msr("classif.acc")))
+```
+
+In this case, there is a clear relationship between the classification accuracy and our custom measure, i.e. the higher the accuracy, the lower the mean absolute error of the predicted probabilities.
+
+```{r}
+cor(score$mae_prob, score$classif.acc)
+```
+
+
+3. "Tune" the `error_train` hyperparameter of the `classif.debug` learner on a continuous interval from 0 to 1, using a simple classification tree as the fallback learner and the penguins task.
+  Tune for 50 iterations using random search and 10-fold cross-validation.
+  Inspect the resulting archive and find out which evaluations resulted in an error, and which did not.
+  Now do the same in the interval 0.3 to 0.7.
+  Are your results surprising?
+
+First, we create the learner that we want to tune, mark the relevant parameter for tuning and set the fallback learner to a classification tree.
+
+```{r}
+lrn_debug = lrn("classif.debug",
+  error_train = to_tune(0, 1),
+  fallback = lrn("classif.rpart")
+)
+lrn_debug
+```
+
+This example is unusual, because we expect better results from the fallback classification tree than from the primary debug learner, which predicts the mode of the target distribution.
```

Comment:
Should this come later? Seems to talk about the results before the code is given.